src/couch_replicator_connection.erl

```diff
@@ -0,0 +1,183 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(couch_replicator_connection).
+
+-behavior(gen_server).
+-behavior(config_listener).
+
+-export([start_link/0]).
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2]).
+-export([code_change/3, terminate/2]).
+
+-export([acquire/1, relinquish/1]).
+
+-export([handle_config_change/5, handle_config_terminate/3]).
+
+-define(DEFAULT_CLOSE_INTERVAL, 60000).
+-define(DEFAULT_CLOSE_BATCH_SIZE, 50).
+
+-record(state, {
+    close_interval,
+    timer,
+    waiting
+}).
+
+-record(connection, {
+    host,
+    port,
+    worker,
+    owner
+}).
+
+-include_lib("ibrowse/include/ibrowse.hrl").
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+
+init([]) ->
+    ?MODULE = ets:new(?MODULE, [named_table, public, {keypos, #connection.worker}]),
+    ok = config:listen_for_changes(?MODULE, self()),
+    Interval = config:get_integer("replicator", "connection_close_interval", ?DEFAULT_CLOSE_INTERVAL),
+    {ok, Timer} = timer:send_after(Interval, close_idle_connections),
+    process_flag(trap_exit, true),
+    {ok, #state{close_interval=Interval, timer=Timer}}.
+
+
+acquire(URL) ->
+    case gen_server:call(?MODULE, {acquire, URL}) of
+        {ok, Worker} ->
+            link(Worker),
+            {ok, Worker};
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+
+relinquish(Worker) ->
+    gen_server:cast(?MODULE, {relinquish, Worker}).
+
+create_worker_async(URL) ->
+    spawn_link(fun() ->
+        {ok, Pid} = ibrowse:spawn_link_worker_process(URL),
+        exit({ok, Pid})
+    end).
+
+handle_call({acquire, URL}, From, State) ->
+    {Pid, _Ref} = From,
+    monitor(process, Pid),
+    case ibrowse_lib:parse_url(URL) of
+        #url{host=Host, port=Port} ->
```

Comment:
Wonder if host includes user and password? Otherwise it would mean sharing connections for  `https://user1:pass1@somecluster.com/db1` and `https://user2:pass2@somecluster.com/db2`?